src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -310,29 +316,53 @@ public CachingRebufferer(ChunkReader file)
         }
 
         @Override
-        public Buffer rebuffer(long position)
+        public BufferHolder rebuffer(long position)
         {
-            int spin = 0;
             try
             {
                 long pageAlignedPos = position & alignmentMask;
-                Buffer buf;
-                Key key = new Key(source, internedPath, pageAlignedPos);
+                BufferHolder buf;
+                Key chunkKey = new Key(source, internedPath, pageAlignedPos);
+
+                int spin = 0;
+                //There is a small window when a released buffer/invalidated chunk
+                //is still in the cache. In this case it will return null
+                //so we spin loop while waiting for the cache to re-populate
                 while (true)
                 {
-                    buf = cache.get(key).reference();
+                    Buffer chunk;
+                    // Using cache.get(k, compute) results in lots of allocation, rather risk the unlikely race...
+                    CompletableFuture<Buffer> cachedValue = cache.getIfPresent(chunkKey);
+                    if (cachedValue == null)
+                    {
+                        // this blocking read might compete with an another read, but the race is benign (under the assumption that caching is valid)
+                        CompletableFuture<Buffer> entry = new CompletableFuture<>();
+                        // Put the future in the cache first so the window of racing (and loading the same chunk twice) is small
+                        cache.put(chunkKey, entry);
+
+                        try
+                        {
+                            chunk = load(chunkKey);
```

Comment:
I am thinking more to the race we have here.
It is possible that 2 threads do the load() and take a ByteBuffer from the pool, but one of them will be leaked (never returned to the pool)

Reply:
I have fixed this problem and added a test (that reproduced the leak and now the leak is no more present)